server/src/main/java/org/elasticsearch/discovery/PeerFinder.java

```diff
@@ -122,35 +122,46 @@ public void activate(final DiscoveryNodes lastAcceptedNodes) {
         logger.trace("activating with {}", lastAcceptedNodes);
 
         synchronized (mutex) {
-            assert assertInactiveWithNoKnownPeers();
+            assert assertInactiveWithNoUndiscoveredPeers();
             active = true;
             activatedAtMillis = transportService.getThreadPool().relativeTimeInMillis();
             this.lastAcceptedNodes = lastAcceptedNodes;
             leader = Optional.empty();
-            handleWakeUp(); // return value discarded: there are no known peers, so none can be disconnected
+            handleWakeUp();
         }
 
         onFoundPeersUpdated(); // trigger a check for a quorum already
     }
 
     public void deactivate(DiscoveryNode leader) {
-        final boolean peersRemoved;
+        final boolean hasInactivePeers;
         final Collection<Releasable> connectionReferences;
         synchronized (mutex) {
             logger.trace("deactivating and setting leader to {}", leader);
             active = false;
             connectionReferences = new ArrayList<>(peersByAddress.size());
-            for (Peer peer : peersByAddress.values()) {
-                connectionReferences.add(peer.getConnectionReference());
+            hasInactivePeers = peersByAddress.isEmpty() == false;
+            final var iterator = peersByAddress.values().iterator();
+            while (iterator.hasNext()) {
+                final var peer = iterator.next();
+                if (peer.getDiscoveryNode() == null) {
+                    connectionReferences.add(peer.getConnectionReference());
+                    iterator.remove();
+                }
             }
-            peersRemoved = handleWakeUp();
+            handleWakeUp();
```

Comment:
This looks like a noop to me now, since all `Peer::handleWakeup` will return false when `active==false`?

Reply:
Good point, yes we can remove that. See 9611859d0a8bf4896c0c97e30ab3aeb241d62a4a.